src/Microsoft.AspNet.Server.Kestrel/Filter/StreamSocketOutput.cs

```diff
@@ -49,8 +49,41 @@ public void Write(ArraySegment<byte> buffer, bool immediate, bool chunk)
 
         public Task WriteAsync(ArraySegment<byte> buffer, bool immediate, bool chunk, CancellationToken cancellationToken)
         {
-            // TODO: Use _outputStream.WriteAsync
-            Write(buffer, immediate, chunk);
+            lock (_writeLock)
+            {
+                if (chunk && buffer.Array != null)
+                {
+                    return WriteAsyncChunked(buffer, cancellationToken);
+                }
+
+                return _outputStream.WriteAsync(buffer.Array ?? _nullBuffer, buffer.Offset, buffer.Count, cancellationToken);
+            }
+        }
+
+        private Task WriteAsyncChunked(ArraySegment<byte> buffer, CancellationToken cancellationToken)
+        {
+            var beginChunkBytes = ChunkWriter.BeginChunkBytes(buffer.Count);
+
+            // When calls return the tasks they have copied the data to the backing buffer
+            // so is safe to send one after another and await at end
+
+            var preambleTask = _outputStream.WriteAsync(beginChunkBytes.Array, beginChunkBytes.Offset, beginChunkBytes.Count, cancellationToken);
+            var dataTask = _outputStream.WriteAsync(buffer.Array ?? _nullBuffer, buffer.Offset, buffer.Count, cancellationToken);
+            var endTask = _outputStream.WriteAsync(_endChunkBytes, 0, _endChunkBytes.Length, cancellationToken);
+
+            if (!preambleTask.IsCompleted)
+            {
+                return Task.WhenAll(preambleTask, dataTask, endTask);
+            }
+            if (!dataTask.IsCompleted)
+            {
+                return Task.WhenAll(dataTask, endTask);
+            }
+            if (!endTask.IsCompleted)
+            {
+                return endTask;
```

Comment:
Is it wrong to simply return `endTask` if it's not completed? Will it ever complete prior to `preambleTask` or `dataTask`?